### DISCUSSION TITLE
read on failed websocket connection error panic

### DISCUSSION THREAD

**Discussion by onur-yildirim-infinitusai** [deepgram/community#643] [General help] [answered]
### Which Deepgram product are you using?

Deepgram SDKs

### Details

I've been running some benchmarking tests locally using the `deepgram-go-sdk@v1.2.0` and consistently encountered an error. Basically, I'm using a couple of audio files to simulate live calls and get STT results. The benchmarking logic sets up a client for each audio file, sends the audio data by splitting it into audio buffers, and then writes the transcript/interim results and latency data to a file.

Here's the issue: after the first file finishes streaming, the other streams keep going for about 9 minutes and then abruptly end with a repeated panic `read on failed websocket connection error.`

I've thoroughly reviewed how I'm initializing the clients, and it seems to work perfectly fine for multiple audios, as long as the difference in audio lengths is less than 9 minutes.

Attached the requestIDs below, `17f64ce1-4b79-4db7-998e-0590145768ad` is the one that I saw the panic

```
panic: repeated read on failed websocket connection

goroutine 725 [running]:
github.com/dvonthenen/websocket.(*Conn).NextReader(0xc001c966e0)
	/Users/onuryildirim/go/pkg/mod/github.com/dvonthenen/websocket@v1.5.1-dyv.2/conn.go:1030 +0x2a8
github.com/dvonthenen/websocket.(*Conn).ReadMessage(0xc0018e3200?)
	/Users/onuryildirim/go/pkg/mod/github.com/dvonthenen/websocket@v1.5.1-dyv.2/conn.go:1093 +0x19
github.com/deepgram/deepgram-go-sdk/pkg/client/live.(*Client).listen(0xc0018e3200)
	/Users/onuryildirim/go/pkg/mod/github.com/deepgram/deepgram-go-sdk@v1.2.0/pkg/client/live/client.go:235 +0x28e
created by github.com/deepgram/deepgram-go-sdk/pkg/client/live.(*Client).ConnectWithRetry
	/Users/onuryildirim/go/pkg/mod/github.com/deepgram/deepgram-go-sdk@v1.2.0/pkg/client/live/client.go:196 +0xa4a
exit status 2
```

This is how I initialize the parallel streams

```
func startBenchmarkingWithDeepgram(firestoreClient *firestoreInternal.Client, allAudioFiles []string) {
	log.Printf("starting benchmarking with deepgram")
	var wg sync.WaitGroup
	for _, audioFile := range allAudioFiles {
		wg.Add(1)
		audioFile := audioFile
		go func() {
			defer wg.Done()
			benchmarkConf := &benchmarkConfiguration{
				vendorName:  "DEEPGRAM",
				audioFile:   audioFile,
				speechCh:    make(chan *sp2txt.STTNBestResult, 4),
				audioCh:     make(chan []byte, inputAudioPacketsBufferSize),
				callPhaseCh: make(chan abspb.CallState_Phase, 4),
			}
			benchmarkForConfiguration(benchmarkConf, firestoreClient)
		}()
	}
	wg.Wait()
}
func benchmarkForConfiguration(benchmarkConfig *benchmarkConfiguration, fs firestoreInternal.InfinitusFirestoreClient) {
	appCtx := loop.NewContext(context.Background(), benchmarkConfig.audioFile)
	// Code for initializing services etc. 

        //picking STT Client based on the benchmarkConfig
	sp2txtStream := getSTTClient(appCtx, t, callService, callKey, benchmarkConfig, fs)
	go runAudioAndClose(appCtx, benchmarkConfig, stream)

	sp2txtStream.StartData(benchmarkConfig.audioCh, benchmarkConfig.callPhaseCh)
	go collectSTTInfo(appCtx, benchmarkConfig)
	appCtx.WaitForDone(func(e *abspb.LoopError) {
		log.Printf("done - %v", e)
	})
	writeLatencyCSV(benchmarkConfig, sp2txtStream.GetLatencyMetrics())
}

func runAudioAndClose(appCtx loop.Context, config *benchmarkConfiguration, stream speechpb.Speech_StreamingRecognizeClient) {
	// Sleep time because we want the STT to start before we send audio
	time.Sleep(1 * time.Second)
	log.Printf("running audio generation: %s", config.audioFile)
	runAudio(config.audioFile, config.audioCh)
	runAudio(benchmarkSilentFile, config.audioCh)
	log.Printf("closing audio channel: %s", config.audioFile)
	close(config.audioCh)
         // This triggers loopContext.WaitForDone
	appCtx.SignalError(errs.NewLoopError(errs.ClassNexmo, errors.New("Websocket Termination"))) 
}
```

### If you are making a request to the Deepgram API and have a request ID, please paste it below:

d2babc78-a16a-41d0-b898-054a18568e61 , e7ec263e-f24d-4611-93d7-59eba22e3c47, 17f64ce1-4b79-4db7-998e-0590145768ad

**Comment by team-deepgram** [comment]
Thanks for asking your question about Deepgram! If you didn't already include it in your post, please be sure to add as much detail as possible so we can assist you efficiently, such as:
- The `request_id` if you have a question about your requests or transcription responses.
- The features you used or the full `api.deepgram.com` URL you sent your request to, including parameters.
- Any code snippets you can share.

**Comment by jkroll-deepgram** [comment, marked as answer]
Hi @onur-yildirim-infinitusai, sorry for the trouble as you're testing our streaming API.

Thank you for providing the request IDs, that gives us a great starting point to dig into your issue.

On several of those requests, I see a log message: `No Websocket messages from client in 12 seconds. Killing connection. Check client connectivity.`

Note that Deepgram closes websockets when it detects that it hasn't received any data from the client in 12 seconds. To avoid this, you can send KeepAlive messages regularly to sustain the connection ([doc link](https://developers.deepgram.com/reference/listen-live#stream-keepalive)).

With the Go SDK, we provide the option to simply add the option of `EnableKeepAlive: true` ([example code link](https://github.com/deepgram/deepgram-go-sdk/blob/60f860ae57b6d10a01be9cca27521739e796ff53/examples/streaming/microphone/main.go#L112-L114)).

Can you please give that a try and see if it resolves your issue?

**Reply by onur-yildirim-infinitusai** [reply]
I wasnt closing the Deepgram connection that are running in parallel correctly so the one terminated one was panicking after some time. Thanks for your help!
